src/dbnode/storage/index_test.go

```diff
@@ -378,6 +381,65 @@ func TestNamespaceIndexQueryNoMatchingBlocks(t *testing.T) {
 	require.NoError(t, err)
 }
 
+func TestNamespaceIndexQueryTimeout(t *testing.T) {
+	ctrl := gomock.NewController(xtest.Reporter{T: t})
+	defer ctrl.Finish()
+
+	test := newTestIndex(t, ctrl)
+
+	now := time.Now().Truncate(test.indexBlockSize)
+	query := index.Query{Query: idx.NewTermQuery([]byte("foo"), []byte("bar"))}
+	idx := test.index.(*nsIndex)
+	idx.state.runtimeOpts.defaultQueryTimeout = time.Second * 1
+
+	defer func() {
+		require.NoError(t, idx.Close())
+	}()
+
+	mockBlock := index.NewMockBlock(ctrl)
+	mockBlock.EXPECT().Stats(gomock.Any()).Return(nil).AnyTimes()
+	blockTime := now.Add(-1 * test.indexBlockSize)
+	mockBlock.EXPECT().StartTime().Return(blockTime).AnyTimes()
+	mockBlock.EXPECT().EndTime().Return(blockTime.Add(test.indexBlockSize)).AnyTimes()
+	mockBlock.EXPECT().
+		Query(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+		DoAndReturn(func(
+			ctx context.Context,
+			c *xresource.CancellableLifetime,
+			q index.Query,
+			opts index.QueryOptions,
+			r index.QueryResults,
+			logFields []opentracinglog.Field,
+		) (bool, error) {
+			for {
+				if !c.TryCheckout() {
+					return false, index.ErrCancelledQuery
+				}
+				time.Sleep(time.Millisecond)
```

Comment:
why is this needed?

Reply:
figured it was better than tight looping in a test.

Reply:
this runs in a separate goroutine (from the worker pool), so probably best to not crush the cpu.